### PR TITLE
[Product CRUD] Full Creation & Updating (Abstract)

### PR DESCRIPTION
**Pull request by emdashcodes**
@mikejolley this PR should be reviewed before #12108 (which will need to be rebased anyway).

It cleans up a little of the code, and finishes implementing create and update, along with some tests.

### CODE REVIEW

File: includes/abstracts/abstract-wc-legacy-product.php [hunk 1]
```diff
@@ -129,6 +129,7 @@ public function get_post_data() {
 	 * @return string
 	 */
 	public function get_title() {
```
Comment (mikejolley) [inline includes/abstracts/abstract-wc-legacy-product.php]: No need to adjust anything yet, but we may want to keep something like this around within the classes for things like simple products which append the parent grouped product name to the title before display. Unless we write a template function to replace this?

File: includes/abstracts/abstract-wc-product.php [hunk 2]
```diff
@@ -477,6 +479,30 @@ public function get_menu_order() {
 		return $this->data['menu_order'];
 	}
 
+	/**
+	 * Returns the product categories. @todo store in class and save?
```
Comment (mikejolley) [inline includes/abstracts/abstract-wc-product.php]: Adding this todo to the main PR checklist.

File: includes/abstracts/abstract-wc-product.php [hunk 3]
```diff
@@ -1004,14 +987,17 @@ public function create() {
...
-			'post_type'     => 'product',
-			'post_status'   => 'publish',
-			'post_author'   => get_current_user_id(),
-			'post_title'    => $this->get_code(),
-			'post_content'  => '',
-			'post_excerpt'  => $this->get_description(),
-			'post_date'     => date( 'Y-m-d H:i:s', $this->get_date_created() ),
-			'post_date_gmt' => get_gmt_from_date( date( 'Y-m-d H:i:s', $this->get_date_created() ) ),
+			'post_type'      => 'product',
+			'post_status'    => 'publish',
```
Comment (mikejolley) [inline includes/abstracts/abstract-wc-product.php]: This needs to be managed within the class. Default can be publish?

File: includes/abstracts/abstract-wc-product.php [hunk 4]
```diff
@@ -1028,7 +1014,19 @@ public function create() {
...
-
+		$post_data = array(
+			'ID'             => $this->get_id(),
+			'post_content'   => $this->get_description(),
+			'post_excerpt'   => $this->get_short_description(),
+			'post_title'     => $this->get_name(),
+			'post_parent'    => $this->get_parent_id(),
+			'comment_status' => $this->get_reviews_allowed(),
+			'menu_order'     => $this->get_menu_order(),
+		);
```
Comment (mikejolley) [inline includes/abstracts/abstract-wc-product.php]: Status update here too.

File: includes/abstracts/abstract-wc-product.php [hunk 5]
```diff
@@ -1062,7 +1060,31 @@ public function delete() {
 	 * @param int $id Object ID.
 	 */
 	private function update_post_meta( $id ) {
-		// update_post_meta( $id, 'discount_type', $this->get_discount_type() );
+		update_post_meta( $id, '_featured', $this->get_featured() );
+		update_post_meta( $id, '_visibility', $this->get_catalog_visibility() );
+		update_post_meta( $id, '_sku', $this->get_sku() );
+		update_post_meta( $id, '_regular_price', $this->get_regular_price() );
+		update_post_meta( $id, '_sale_price', $this->get_sale_price() );
+		update_post_meta( $id, '_regular_price', $this->get_regular_price() );
```
Comment (mikejolley) [inline includes/abstracts/abstract-wc-product.php]: This is duplicated.

**Review by mikejolley — COMMENTED** [review]
Added my notes 👍

**Comment by justinshreve** [comment]
@mikejolley Thanks! Updated & ready for a second look before merge.

**Pull request merged**